fedimint-core/src/task.rs

```diff
@@ -173,27 +174,55 @@ impl TaskGroup {
         let handle = self.make_handle();
 
         let (tx, rx) = oneshot::channel();
-        let handle = crate::runtime::spawn(&name, {
-            let name = name.clone();
-            async move {
-                // Unfortunately log levels need to be static
-                if quiet {
-                    trace!(target: LOG_TASK, "Starting task {name}");
-                } else {
-                    debug!(target: LOG_TASK, "Starting task {name}");
-                }
-                // if receiver is not interested, just drop the message
-                let r = f(handle).await;
-                if quiet {
-                    trace!(target: LOG_TASK, "Finished task {name}");
-                } else {
-                    debug!(target: LOG_TASK, "Finished task {name}");
-                }
-                let _ = tx.send(r);
-            }
-        });
-        self.inner.add_join_handle(name, handle);
-        guard.completed = true;
+        self.inner
+            .active_tasks_join_handles
+            .lock()
+            .expect("Locking failed")
+            .insert_with_key(move |task_key| {
+                (
+                    name.clone(),
+                    crate::runtime::spawn(&name, {
+                        let name = name.clone();
+                        async move {
+                            defer! {
+                                // Panic or normal completion, it means the task
+                                // is complete, and does not need to be shutdown
+                                // via join handle. This prevents buildup of task
+                                // handles.
+                                if handle
```

Comment:
And removes itself from it on drop.

Thanks to that, we don't need to wait or even remember about it anymore.